xbmc/pvr/channels/PVRChannelProviders.cpp

```diff
@@ -0,0 +1,370 @@
+/*
+ *  Copyright (C) 2012-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "PVRChannelProviders.h"
+
+#include "ServiceBroker.h"
+#include "pvr/PVRDatabase.h"
+#include "pvr/PVRManager.h"
+#include "pvr/addons/PVRClient.h"
+#include "pvr/addons/PVRClients.h"
+#include "pvr/channels/PVRChannelProvider.h"
+#include "settings/Settings.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+
+#include <algorithm>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace PVR;
+
+bool CPVRChannelProvidersContainer::UpdateFromClient(
+    const std::shared_ptr<CPVRChannelProvider>& provider)
+{
+  CSingleLock lock(m_critSection);
+  std::shared_ptr<CPVRChannelProvider> providerToUpdate =
+      GetByClient(provider->GetClientId(), provider->GetUniqueId());
+  if (providerToUpdate)
+  {
+    return providerToUpdate->UpdateEntry(provider, ProviderUpdateMode::BY_CLIENT);
+  }
+  else
+  {
+    provider->SetProviderId(++m_iLastId);
+    InsertEntry(provider, ProviderUpdateMode::BY_CLIENT);
+  }
+
+  return true;
+}
+
+std::shared_ptr<CPVRChannelProvider> CPVRChannelProvidersContainer::GetByClient(int iClientId,
+                                                                                int iUniqueId) const
+{
+  CSingleLock lock(m_critSection);
+  for (const auto& provider : m_providers)
+  {
+    if (provider->GetClientId() == iClientId && provider->GetUniqueId() == iUniqueId)
+    {
+      return provider;
+    }
+  }
+
+  return std::shared_ptr<CPVRChannelProvider>();
+}
+
+void CPVRChannelProvidersContainer::InsertEntry(
+    const std::shared_ptr<CPVRChannelProvider>& newProvider, ProviderUpdateMode updateMode)
+{
+  bool found = false;
+  for (auto& provider : m_providers)
+  {
+    if (provider->GetClientId() == newProvider->GetClientId() &&
+        provider->GetUniqueId() == newProvider->GetUniqueId())
+    {
+      found = true;
+      provider->UpdateEntry(newProvider, updateMode);
+    }
+  }
+
+  if (!found)
+  {
+    m_providers.emplace_back(newProvider);
+  }
+}
+
+CPVRChannelProviders::CPVRChannelProviders()
+{
+}
```

Comment:
Remove here and define in the header as `= default`if you like or remove it even there.

Reply:
done